doc/ug/dv_methodology.md

```diff
@@ -299,65 +299,124 @@ Once the regression tool is developed, it will provide a way to enable these cap
+These constructs also allow us to capture whether multiple scenarios have occurred simultaneously through crosses.
+This is also often referred to as cross coverage.
+These constructs are typically encapsulated in a class and are sometimes referred to as the ‘functional coverage model’.
+They are sampled in ‘reactive’ components of the testbench, such as monitors and/or the scoreboards.
 They can also be embedded within the RTL to sample interesting scenarios through DV stimulus.
 
 Here are the metrics used with a brief explanation:
 
-*  **Assert Coverage**: This metric measures which assertions, cover properties and sequences have been exercised in the course of the simulation.
+* **Assert Coverage**: This metric measures which assertions, cover properties, and sequences have been exercised in the course of the simulation.
   Note, an assertion precondition counts as a cover point.
-*  **Covergroup Coverage**: This metric (sometimes called "Testbench Coverage") measures covergroups during simulation.
-  These are usually coded in the testbench to check that the desired stimulus was generated properly, but can also be embedded in the RTL.
-
-Code coverage can reach 100% but it may not be sufficient to indicate whether all interesting combinations of scenarios were exercised (such as boundary conditions, concurrent scenarios and different CSR fields holding specific combinations of values resulting in the DUT being exercised in interesting ways).
-Hence, developing a detailed functional coverage model in conjunction with constrained random test sequences is key to high quality DV effort.
-
-Functional coverage can also reach 100% while the code coverage may not.
-This would indicate that the testplan is either incomplete or there are undocumented or unreachable features in the design.
-As such they can be addressed accordingly through either updating the design specification and augmenting the testplan / written tests, or by optimizing the design to remove unreachable logic if possible.
-There may be features that cannot be tested and cannot be removed from the design either.
-These would have to be analyzed and excluded from the coverage collection as a waiver.
-This is an exercise the DV and the designer typically perform together.
-This is discussed in more detail below.
+* **Cover group Coverage**:  Observes values on busses, registers, etc.
+  can verify that a specific value or a range of values was seen on a bus.
+  And that no illegal values were seen.
+  Could also be the intersection of two cover points.
+  One example could be observing the FIFO full signal and the FIFO write enable signal is asserted at the same time.
+  This is called a coverage cross and will give valuable information on whether this scenario was ever seen. 
+* **Cover Property coverage**: This observes events or series of events.
```

Comment:
is this same as `Assert Coverage`

Reply:
@weicaiyang 
ahh yes you are correct - this is a slip in the copy paste.
what I meant to put here is SVA systemverlog assertions which are not the same as cover property/assert coverage.

I like the name cover property better than assert coverage. because assert only describes the assert part of the procedural statement and not the cover part.
i.e 
you can have both or either of the 
```
assert (a)
cover (a)
```
in your procedural cover statement.

would this satisfy you?

Reply:
these 2 are all SVA syntax. Both are used along with design or interface. Not sure if we need to separate them.

Reply:
@weicaiyang 
I always mentally separated the two methods as SVA for internal functional coverage.
and cover property for something that can be observed on the pins, and at the same time
SVA as the immediate assertion and cover property as the concurrent assertions with using the keywords
``
sequence 
property
```


I never really looked into the actual definition of the two.
which I did now.

so it is possible to write a complex property cover point using procedural code only, and for the very complex ones, I think this is probably the best choice.
but I practice I think you are correct, SVA is usually the choice as they are simpler.

I will merge the two.

Reply:
i.e cover property is the coverage of temporal relations. where SVA is just a tool we can use to cover them.

Reply:
@weicaiyang I deleted my response as I was answering a different question.
this is a really good comment, really made me think and I changed my mind several times

as a start I would like to say - yes you are correct property coverage is usually done with system verilog assertions.

I somehow made up my mind (without checking) that immediate assertions was assertions and concurrent/temporal assertions were property coverage. which is not completely true.
assertions are just a tool that can be used to collect coverage of temporal relationships between signals.

but back to the document.
I actually distinguish between the two.
assertion coverage - check which assertions have fired.
cover property (using SVA) check temporal relations of signals

I am not sure this warrants having them both. I am trying to think if they can give different information.
can we store this for a later point when I get to the examples?
if you prefer I can merge them now and split them again later if I find it useful?

Reply:
I feel we may categorize them like this 
- assertion coverage
   - assert property coverage
   - cover property coverage
- covergroup coverage

Reply:
here my problem.
I don't believe they are the same. 
The documentation I find speak of them as different.

"systemverilog for verification - by Chris spear / Greg Tumbush" 

> There is a large overlap between sequences that are coded procedurally or using SVA. Assertions can have local variables and perform simple data checking. If you need to check a more complex protocol, such as determining whether a packet successfully went through a router, procedural code is often better suited for the job.

The Mentor coverage user guide also lists them as two different properties.
cover properties are also mentioned separately in the IEEE systemverilog specification. #16.8

So I am really trying to educate my self here.
I never actually seen any verification implemented with sequences and cover properties.
I always personally used SVA to cover this. ( which is why I also claimed "usually" in my comment before)

how about this.

- Cover Property Modelling 
  - systemverilog assertions
  - procedural code

I have updated the code to my suggestion - let me know what you think
@sriyerg @rswarbrick 
do you have any input on this?

Reply:
the change looks good to me. Thanks